### PR TITLE
Fix completions of optional properties in generic positions

### LINKED ISSUE #28470: Intellisense missing for generic type extending an all partial type
**TypeScript Version:**
3.1.3 and 3.2.0-dev.20181110

**Search Terms:**
Intellisense suggestions generic type

**Code**

```ts
interface Options {
    someFunction?: () => string
    anotherFunction?: () => string
}

export class Clazz<T extends Options> {
    constructor(public a: T) {
    }
}

new Clazz({

// ctrl + space on above line does not provide any suggestion
// it does when:
// - making some/anotherFunction required instead of optional (in Options)
//   OR
// - giving constructor's first param a type of `T | Options`
})
```

**Expected behavior:**
suggest `someFunction` and `anotherFunction`.

**Actual behavior:**
no suggestions

**Playground Link:** 
http://www.typescriptlang.org/play/#src=interface%20Options%20%7B%0D%0A%20%20%20%20someFunction%3F%3A%20()%20%3D%3E%20string%0D%0A%20%20%20%20anotherFunction%3F%3A%20()%20%3D%3E%20string%0D%0A%7D%0D%0A%0D%0Aexport%20class%20Clazz%3CT%20extends%20Options%3E%20%7B%0D%0A%20%20%20%20constructor(public%20a%3A%20T)%20%7B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Anew%20Clazz(%7B%0D%0A%0D%0A%2F%2F%20ctrl%20%2B%20space%20in%20line%2012%20does%20not%20provide%20any%20suggestion%0D%0A%2F%2F%20it%20does%20when%3A%0D%0A%2F%2F%20-%20making%20some%2FanotherFunction%20required%20instead%20of%20optional%20(in%20Options)%0D%0A%2F%2F%20%20%20OR%0D%0A%2F%2F%20-%20giving%20constructor's%20first%20param%20a%20type%20of%20%60T%20%7C%20Options%60%0D%0A%7D)%0D%0A

**Related Issues:**

### LINKED ISSUE #30507: Autocompletion broken for generic function parameter with optional properties
As #30212 has been closed incorrectly (it's not a duplicate) and the comments there are not reacted on anymore, I reopen the issue here.

- VSCode Version: Version 1.31.1 (1.31.1)
- OS Version: macOS High Sierra 10.13.6

Steps to Reproduce:

typescript autocompletion

Open a TypeScript file with this content:

```ts
type Params = {
  foo?: string
  bar?: number
  visible?: boolean
}

export const obj = {
  paramsIdentityInline<T extends Params>(params: T): T {
    return params
  },
}

obj.paramsIdentityInline({
  
})

function direct(params: Params) {}

direct({

})
```

When you have a type which only consists of optional properties and use it in a generic context as a type parameter, the autocompletion doesn't work, while it works when you use that type directly without being generic in the function signature.
However, as soon as at least one of the properties on that type is non-optional, the autocompletion works, even for the optional properties.

Does this issue occur when all extensions are disabled?: Yes

As @hediet correctly notes, the referenced fix doesn't work for this bug.
Please see #30212 for more infos.

Ping @mjbvz

### PR DESCRIPTION
**Pull request by andrewbranch**
Builds on #32100

Fixes #30507
Fixes #28470

Notable changes from #32100:

- Doesn’t cache an alternative type during inference (@weswigham’s change request on the original PR)
- Rather than getting the contextual type from the base signature alone, this intersects the types from the base signature and the resolved signature, so that in a case like

```ts
type U = { x: boolean };
declare function f<T extends Partial<U>>(x: T, y: T)

f({ a: true }, { /**/ });
```

you get completions for both `x` _and_ `a`. Thanks to @DanielRosenwasser for that suggestion.

### CODE REVIEW

**Comment by timsuchanek** [comment]
Awesome! Would be great to see this merged!

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20887,19 +20881,23 @@ namespace ts {
...
-            return argIndex === -1 ? undefined : getContextualTypeForArgumentAtIndex(callTarget, argIndex);
+            return argIndex === -1 ? undefined : getContextualTypeForArgumentAtIndex(callTarget, argIndex, contextFlags);
         }
 
-        function getContextualTypeForArgumentAtIndex(callTarget: CallLikeExpression, argIndex: number): Type {
+        function getContextualTypeForArgumentAtIndex(callTarget: CallLikeExpression, argIndex: number, contextFlags?: ContextFlags): Type {
             // If we're already in the process of resolving the given signature, don't resolve again as
             // that could cause infinite recursion. Instead, return anySignature.
-            const signature = getNodeLinks(callTarget).resolvedSignature === resolvingSignature ? resolvingSignature : getResolvedSignature(callTarget);
+            const signature = getNodeLinks(callTarget).resolvedSignature === resolvingSignature ? resolvingSignature : getResolvedSignature(callTarget, /*candidatesOutArray*/ undefined, /*checkMode*/ undefined);
```
Comment (sandersn) [inline src/compiler/checker.ts]: Adding two `undefined`s here should be unnecessary.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -20887,19 +20881,23 @@ namespace ts {
...
             if (isJsxOpeningLikeElement(callTarget) && argIndex === 0) {
                 return getEffectiveFirstArgumentForJsxSignature(signature, callTarget);
             }
+            if (contextFlags && contextFlags & ContextFlags.Completion && signature.target) {
+                const baseSiganture = getBaseSignature(signature.target);
```
Comment (sandersn) [inline src/compiler/checker.ts]:
```suggestion
                const baseSignature = getBaseSignature(signature.target);
```

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -20887,19 +20881,23 @@ namespace ts {
...
+                return intersectTypes(getTypeAtPosition(signature, argIndex), getTypeAtPosition(baseSiganture, argIndex));
```
Comment (sandersn) [inline src/compiler/checker.ts]:
```suggestion
                return intersectTypes(getTypeAtPosition(signature, argIndex), getTypeAtPosition(baseSignature, argIndex));
```

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -24381,7 +24379,7 @@ namespace ts {
                 return resolveErrorCall(node);
             }
 
-            return resolveCall(node, callSignatures, candidatesOutArray, checkMode, isOptional);
+            return resolveCall(node, callSignatures, candidatesOutArray, checkMode, isOptional, /*fallbackError*/ undefined);
```
Comment (sandersn) [inline src/compiler/checker.ts]: these undefineds shouldn't be needed either

File: tests/cases/fourslash/completionsWithOptionalPropertiesGeneric.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/// <reference path="fourslash.ts" />
+// @strict: true
+
+//// interface MyOptions {
+////     hello?: boolean;
+////     world?: boolean;
+//// }
+//// declare function bar<T extends MyOptions>(options?: Partial<T>): void;
```
Comment (sandersn) [inline tests/cases/fourslash/completionsWithOptionalPropertiesGeneric.ts]: mind adding a case with unconstrained `T`? It should have no completions I think, but I just want to make sure.

File: src/compiler/checker.ts (hunk 2 above)
Comment (orta) [reply src/compiler/checker.ts]: I use (and love) https://github.com/streetsidesoftware/vscode-spell-checker

It runs over the website now: https://github.com/microsoft/TypeScript-Website/pull/84#issuecomment-541370379

**Comment by AviVahl** [comment]
@andrewbranch this also fixes https://github.com/microsoft/TypeScript/issues/28470

File: src/compiler/checker.ts (hunk 4 above)
Comment (andrewbranch) [reply src/compiler/checker.ts]: Ah yeah, forgot to delete those when removing the extra param added in the original PR.

```suggestion
            return resolveCall(node, callSignatures, candidatesOutArray, checkMode, isOptional);
```

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -24452,7 +24450,7 @@ namespace ts {
                     return resolveErrorCall(node);
                 }
 
-                return resolveCall(node, constructSignatures, candidatesOutArray, checkMode, /*isOptional*/ false);
+                return resolveCall(node, constructSignatures, candidatesOutArray, checkMode, /*isOptional*/ false, /*fallbackError*/ undefined);
```
Comment (andrewbranch) [inline src/compiler/checker.ts]:
```suggestion
                return resolveCall(node, constructSignatures, candidatesOutArray, checkMode, /*isOptional*/ false);
```

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -24461,7 +24459,7 @@ namespace ts {
             // operation is Any. It is an error to have a Void this type.
             const callSignatures = getSignaturesOfType(expressionType, SignatureKind.Call);
             if (callSignatures.length) {
-                const signature = resolveCall(node, callSignatures, candidatesOutArray, checkMode, /*isOptional*/ false);
+                const signature = resolveCall(node, callSignatures, candidatesOutArray, checkMode, /*isOptional*/ false, /*fallbackError*/ undefined);
```
Comment (andrewbranch) [inline src/compiler/checker.ts]:
```suggestion
                const signature = resolveCall(node, callSignatures, candidatesOutArray, checkMode, /*isOptional*/ false);
```

File: src/compiler/checker.ts (hunk 1 above)
Comment (andrewbranch) [reply src/compiler/checker.ts]:
```suggestion
            const signature = getNodeLinks(callTarget).resolvedSignature === resolvingSignature ? resolvingSignature : getResolvedSignature(callTarget);
```

**Comment by andrewbranch** [comment]
Ugh, committing PR review suggestions from GitHub UI doesn’t work in checker.ts 😑

**Comment by DanielRosenwasser** [comment]
🎉

**Comment by timsuchanek** [comment]
Very excited to see this merged 🙏 
One question: How can I try it out in VSCode?
Unfortunately I'm not able to use this properly yet, even though setting `"typescript.tsdk"` in my `settings.json` to `node_modules/typescript/lib
`.

This example should now work in VSCode and show `world` as a completion:

```ts
interface MyOptions {
  hello?: boolean
  world?: boolean
}
declare function bar<T extends MyOptions>(options?: Partial<T>): void

bar({
  hello: true,
  /* cursor */
})
```

@andrewbranch can you confirm that this example is now working for you?

**Comment by andrewbranch** [comment]
@timsuchanek yep, that example works in VS Code for me:

image

> setting `"typescript.tsdk"` in my `settings.json` to `node_modules/typescript/lib`.

If you’re building the source from master, you want to set the path to `built/local`. The one in the repo’s node_modules is just whatever the nightly was whenever you last did a clean `npm install` 😄

**Comment by timsuchanek** [comment]
Awesome. I didn't know about the `built/local`. That also works for me.
Thanks a lot!

**Pull request merged**